src/webview/memory-webview-view.tsx

```diff
@@ -135,21 +151,48 @@ class App extends React.Component<{}, MemoryAppState> {
         messenger.onRequest(getWebviewSelectionType, () => this.getWebviewSelection());
         messenger.onNotification(showAdvancedOptionsType, () => this.showAdvancedOptions());
         messenger.sendNotification(readyType, HOST_EXTENSION, undefined);
+        this.updateAutoRefresh();
     }
 
-    public componentDidUpdate(_: {}, prevState: MemoryAppState): void {
-        const addressPaddingNeedsUpdate =
-            (this.state.addressPadding === 'Min' && this.state.memory !== prevState.memory)
-            || this.state.addressPadding !== prevState.addressPadding;
-        if (addressPaddingNeedsUpdate) {
+    public componentDidUpdate(_: {}, from: MemoryAppState): void {
+        const current = this.state;
+        const stateChange: Change<MemoryAppState> = { from, to: current };
+        const sessionContextChange: Change<SessionContext> = { from: from.sessionContext, to: current.sessionContext };
+        const viewStateChange: Change<ViewState> = { from: from.viewState, to: current.viewState };
+
+        if (hasChanged(stateChange, 'addressPadding') || (this.state.addressPadding === 'Minimal' && hasChanged(stateChange, 'memory'))) {
             const effectiveAddressLength = this.getEffectiveAddressLength(this.state.memory);
             if (this.state.effectiveAddressLength !== effectiveAddressLength) {
                 this.setState({ effectiveAddressLength });
             }
         }
+        if (hasChanged(stateChange, 'autoRefresh') || hasChanged(stateChange, 'autoRefreshDelay')) {
+            this.updateAutoRefresh();
+        }
+
+        if (current.autoRefresh === 'On Stop' && hasChangedTo(sessionContextChange, 'stopped', true) ||
+            current.autoRefresh === 'On Focus' && hasChangedTo(viewStateChange, 'active', true)) {
+            this.fetchMemory();
+        }
+
         hoverService.setMemoryState(this.state);
     }
 
+    componentWillUnmount(): void {
+        clearTimeout(this.refreshTimer);
+    }
+
+    protected updateAutoRefresh = (): void => {
+        clearTimeout(this.refreshTimer);
+
+        if (this.state.autoRefresh === 'After Delay' && this.state.autoRefreshDelay && this.state.autoRefreshDelay > 0) {
+            // we do not use an interval here as we only want to schedule another refresh AFTER the previous execution AND the delay has passed
+            // and not strictly every n milliseconds. Even if 'fetchMemory' fails here, we schedule another auto-refresh.
+            const scheduleRefresh = () => this.fetchMemory().finally(() => this.updateAutoRefresh());
```

Comment:
Currently in `fetchMemory`, we are not `await`ing `doFetchMemory` so it will return immediately and we won't have waited for the memory to be fetched.

Reply:
Good catch! I now return the `doFetchMemory` so we are actually working with the correct promise!